README.md
<img align="left" width="175" height="175" src="./doc/images/tropo.png">

```
CAUTION: When running the MTP instrument, if Tsynth under the Engineering tab
gets to 50 degreesC, the probe needs to be shutdown to avoid overheating
```
# Microwave Temperature Profiler Instrument Control, Real-Time Display and Processing Software

The Microwave Temperature Profiler (MTP) is a HAIS instrument which was developed by JPL for the NSF/NCAR GV.

The [original VB6 code base](https://github.com/NCAR/MTP-VB6) from JPL that supports data processing and display was written in Microsoft Visual Basic 6. Since Microsoft has chosen to depracate Visual Basic, RAF has embarked on a re-write of some components of the original code. This repository contains those rewrites.

Packages needed to run the MTP software are documented in mtpenv.yml

## To install (on MAC/LINUX/Windows10)

Make sure you have installed at least python 3.7
Use miniconda to install all needed packages:
 * Download conda from https://conda.io/miniconda.html and install
   * Update conda if warned following instructions
 * Install Git (if not already there) and download MTP:
   * https://git-scm.com/ -> Download latest per automatic OS detection. Run .exe file to install. I used default settings as suggested by installer, except that I asked to install a desktop icon for “Git Bash”
 * If installing on eol-ale or eol-atom, set up the EOL Jenkins user to download software from github
   * Create C:\Users\lroot\.gitconfig containing
```
[user]
        name = eolJenkins
```
   * Create C:\Users\lroot\.git-credentials containing the eoljenkins git credential - either login to github as user jenkins to get it, or copy it from the other MTP laptop.
```
https://eoljenkins:<long credential string here>@github.com
```
 * Launch “Git Bash”
 * From /c/Users/lroot in git bash
```
    git clone http://github.com/NCAR/MTP
```
   The first time you may need to cancel out of a window asking for credentials before the download will start.
 * Create the MTP conda environment from the YAML env file
```
   > conda env create -f mtpenv.yml
```
 * If the yaml file doesn't work because it can't resolve conflicts, you can install by hand:
```
   > conda create --name mtp
```
   then activate and conda install each package individually. Note that metpy is only available from the conda-force channel.
 * Activate the mtp conda environment - see - https://conda.io/activation
```
   > conda init bash
   > conda activate mtp
```
 * Install the EOL-Python packages per instructions in https://github.com/NCAR/EOL-Python
 
## Optional on Windows10

 * Add Miniconda3 and Miniconda3\condabin to your path
   * Windows search -> type "env" -> click "Edit the system environment variables"
   * In lower "System variables" window, click the "Path" row and click edit
   * Click "New" and add the new paths, e.g.
     * C:\Users\lroot\Miniconda3
     * C:\Users\lroot\Miniconda3\condabin
     * C:\Users\lroot\Miniconda3\Scripts
     * C:\Users\lroot\Miniconda3\Library\bin

If the packages are not available via the conda-forge channel, you can search for alternative channels at https://anaconda.org

Change you environment variable and add a PYTHONPATH that points to the netCDF installation (You will also add the path to EOL-Python to this env var below.)

### Configure the MTPviewer software
 * Create shortcuts to the bat files in the windows10 dir and put them on the Desktop
 * Edit the bat file and correct the python path if needed
 * Add --config=path\to\config\file for project you are going to run.
 * If you want to run the IWGemulator, install aircraft_nc_utils
```
    > git clone http://github.com/NCAR/aircraft_nc_utils
```

Install the EOL-Python packages per instructions in https://github.com/NCAR/EOL-Python

Check your PYTHONPATH
```
    > echo %PYTHONPATH%
```
It should contain a path to EOL-Python and a path to the anaconda site-packages.
If it doesn't follow the instructions on the EOL-Python page.

## To operate the MTP from Windows10

You only need to do this if you will be connecting your computer directly to the MTP instrument in the lab or on the aircraft. If you are only running MTPviewer to monitor collected data, you don't need to install the driver.

 * Install the driver for the USOPTL4 USB to serial converter
 * Download the driver from https://support.advantech.com/support/DownloadSRDetail_New.aspx?SR_ID=1-HIPU-30&Doc_Source=Download 

 * Extract the zip file to the desktop
 * Run C:\Documents and Settings\mtp\Desktop\USB_Drivers_PKG_v2-08-28\BBSmartWorx\Windows\dpinst64
 * Plug in the USB cable from the MTP to the laptop
 * Go to the device manager and change the port to comm 6
   * This PC -> right click -> manage -> device manager -> Other Devices or COM & PORT (I think) -> USOPTL4
 * Comfirm port settings as 9600|8|None|1|None

Information on operating the MTP, and other documentation, can be found on the (UCAR SEW MTP wiki)[https://wiki.ucar.edu/display/SEW/MicrowaveTemperatureProfiler]

## To run this code:
 * Set up a project directory. See: https://github.com/NCAR/MTP/blob/master/config/README.md
 * cd src
 * On Windows10:
 ```
 * click on MTPviewer icon on the desktop. If this is not available:
     > conda activate (to get the base environment where libraries have been installed)
     > C:\Users\lroot\Miniconda3\python.exe MTPviewer.py --config=path\to\config\fie

 ```
 * On a MAC:
 ```
 > python3 MTPviewer.py --config=\path\to\config\file
 ```
** NOTE that on a MAC you will use python3, but on Windows it's python.exe (no 3) **

## To run in test mode, generate fake "real-time" data by running

 * On Windows10:
 ```
 * Click on MTPemulator icon on the desktop. If this is not available:
    > conda activate
    > cd C:\Users\lroot\MTP\src\emulator
    > C:\Users\lroot\Miniconda3\python.exe snd_MTP_udp.py

 * Click on IWGemulator icon on the desktop. If this is not available:
    > git clone http://github.com/NCAR/aircraft_nc2iwg1
    > conda activate
    > cd C:\Users\lroot\MTP\Data\NGV\DEEPWAVE\NG
    > C:\Users\lroot\Miniconda3\python.exe C:\Users\lroot\aircraft_nc2iwg1\nc2iwg1.py -s 1 -u True -er True DEEPWAVERF01.nc

 ```
 * On a MAC:
 ```
> cd MTP/src/emulator
> python3 snd_MTP_udp.py
> ./snd_IWG.sh  (need to install http://github.com/NCAR/aircraft_nc2iwg1)
```
* Then run the GUI in real-time mode, using the platform-specific python call, e.g. on Windows:
```
>python MTPviewer.py --config=\path\to\config\file
```

## Developer Notes

### Documentation

For complete documentation on each class/method, useful if you need to modify the code, use pydoc to extract embedded documentation from each file:

* On a MAC: (Change python3 to python for Windows)
```
> cd src
> python3 -m pydoc <filename>
e.g. python3 -m pydoc lib/rwget.py
```

### Unit tests

If the unittests are all run sequentially from the same command (python3 -m unittest discover -s ../tests -v), earlier tests seem to leave the unittest code in a state that causes subsequent tests to fail. An attempt was made to get each test to clean up after itself by adding setUp and tearDown functions. But Python's unittest holds on to all sorts of memory until the entire test suite has been run. For a good explanation see: https://stackoverflow.com/questions/26915115/unittest-teardown-del-all-attributes/35001389

To get around this, a shell script has been written that breaks up the test suite into smaller chunks. To manually run all unittests, use this script:

* On a MAC: (Change python3 to python for Windows) 
```
> cd src
OR
> cd tests
> ../tests/run_tests.sh
```

* On a WINDOWS10:
```
> cd src
OR
> cd tests
> powershell -noexit "& ""..\tests\run_tests_Windows.ps1"""
```


CHANGELOG.md
# Changelog

Changelog for MTP software

## [Unreleased]

### Added
- Write to `logfile` by default. Set default loglevel to WARNING and add --v command line option to get info level messages sent to console. When --debug called, debug messages go to both console and logfile. In GUI mode, only WARNING and ERROR message generate message display boxes.

- Implement working control program GUI for `MTPcontrol` including display of state info.

- Add a complete sample project setup dir as LAB.

- Add `limit checking` per Josh's recommended ranges.

### Changed
- Update `tests` and `code` to work with new refactored EOLpython logger and to use class variable instead of env var to indicate testing mode. Combine all tests calls under one run_tests script.

- Fix `BUG` in integrator logic that was cause of first integration (Angle 1, Channel 1) being zero during ACCLIP FF03.

- Remove excess checks and adjust timing on some loops to tighten up scan time. Scan time is now ~19 seconds in the lab which is the same as the VB6.

- Update `MTPviewer` to still display counts and housekeeping when IWG is not available. When IWG comes back, may need to restart viewer to see derived temperature profile. (This is a bug I haven't tracked down yet.)

- Disambiguate send_port and read_port now that both MTPcontrol and MTPviewer
get ports from config file.

- MTPviewer: If RCFs are missing (RC dir is empty), or other errors that cause
prfiles not to be generated, code will catch this and display a message on the
profile plot rather than pop up a warning box.

## [0.2] - 2022-08-29 - ACCLIP FF03 TEST

This version of the software was flown on ACCLIP FF03 from Anchorage to RMMA
for additional testing.

### Added

- Implement `IWG averaging`. Average starts with the mirror in the home position,
and accumulates IWG packets thru a complete B line mirror rotation back to home
and continuing through the collection of target and housekeeping data. The data
reported in the IWG line is from the last IWG packet received. Instantaneous IWG
values are still used to do the pointing correction.

- Tighten up `scan time`. With emulator, scan completes in 6 seconds, so I am
hopeful scan will complete in 17 seconds during flight. This will be tested
during FF03.

- Partial implementation of updated `GUI` functionality. Does not work yet.

- Watermark on curtain plot indicating preliminary data if in realtime mode

### Changed

- The number of significant digits reported in the `A line` has been updated to 
better match the Raw files generated by the VB6 code.

- Improved `error handling`

- Default `mtp_emulator` chaos level is now low.

### Removed

- Additional unused files have been removed

## [0.1] - 2022-08-02 - ACCLIP FF02 TEST

This version of the software was flown for ACCIP FF02 leg 2 from Sacramento to
Kona as a first test under flight conditions.

### Added

- Jenkins configuration is saved to a `jenkinsfile`

- Users can now access documentation of the configuration directory
layout and a sample YAML `config file`

- Deploy conda runtime environment via `mtpenv.yml` file

- `mtudpd` now has an option to send to RIC

- A suite of scripts to run the MTP manually for testing

- Initialize the IWG section of the MTP dictionary using the variable list
provided in the ascii_parms file.

- Implement a few unit tests for the control software - incomplete, but
a start.

### Changed

- Fully refactored and updated the `MTPcontrol` program

- Updated deprecated `Oxygen frequencies` to be accurate

- Fixed pointing calculations

- Code now handles varying numbers of flight levels in the RCF file
programmatically.

### Removed

- Unused files have been removed

## [1ba299d] - 2022-03-15

- first tagged release

<!-- Versions -->
[unreleased]: https://github.com/NCAR/MTP/compare/V0.2-ACCLIPff03...HEAD
[0.2]: https://github.com/NCAR/MTP/compare/V0.1-ACCLIPff02...V0.2-ACCLIPff03
[0.1]: https://github.com/NCAR/MTP/compare/pre-tig3er...V0.1-ACCLIPff02
[1ba299d]: https://github.com/NCAR/MTP/releases/tag/pre-tig3er


src/util/MTP.py
###############################################################################
#
# This python dictionary stores the content of a single raw MTP record
#
# A <date> <time> SAPITCH SRPITCH SAROLL SRROLL SAPALT SRPALT SAAT SRAT SALAT
#         SRLAT SALON SRLON SMCMD SMENC
# IWG1,<date>T<time>,...
# B SCNT[30]
# M01: VM08CNTE VVIDCNTE VP08CNTE VMTRCNTE VSYNCNTE VP15CNTE VP05CNTE VM15CNTE
# M02: ACCPCNTE TDATCNTE TMTRCNTE TAIRCNTE TSMPCNTE TPSPCNTE TNCCNTE TSYNCNTE
# Pt: TR350CNTP TTCNTRCNTP TTEDGCNTP TWINCNTP TMIXCNTP TAMPCNTP TNDCNTP
#         TR600CNTP
# E TCNT[6]
#
# Written in Python 3
#
# COPYRIGHT:   University Corporation for Atmospheric Research, 2019
###############################################################################
import numpy

MTPrecord = {
    # A string to hold an Ascii packet before it is parsed
    'asciiPacket': "",

    # A dictionary to hold a single raw data scan. A raw scan consists of
    # a set of 7 different line types
    'Aline': {
        # Regular expressions to match the various scan lines
        're': "^A (........) (..):(..):(..) (.*)",
        'found': False,
        'date': "",  # YYYYMMDDTHHMMSS
        'data': [],  # A string containing the data values after date/time
        'values': {'DATE': {  # MTP Scan Date (YYYYMMDD)
                  'val': numpy.nan},
                 'timestr': {  # MTP Scan Time (HHMMSS)
                  'val': numpy.nan},
                 'TIME': {  # MTP Scan Time (HHMMSS) converted to secs
                  'val': numpy.nan},
                 'SAPITCH': {  # MTP Scan Avg Pitch (degree)
                  'val': numpy.nan, 'idx': 0,
                  'short_name': 'platform_pitch',
                  'units': 'degree',
                  'long_name': 'Aircraft pitch (deg)',
                  '_FillValue': "-99.9"},
                 'SRPITCH': {  # MTP Scan RMSE Pitch (degree)
                  'val': numpy.nan, 'idx': 1},
                 'SAROLL':  {  # MTP Scan Avg Roll (degree)
                  'val': numpy.nan, 'idx': 2,
                  'short_name': 'platform_roll',
                  'units': 'degree',
                  'long_name': 'Aircraft roll (deg)',
                  '_FillValue': "-99.9"},
                 'SRROLL':  {  # MTP Scan RMSE Roll (degree)
                  'val': numpy.nan, 'idx': 3},
                 'SAPALT':  {  # MTP Scan Avg Pressure Altitude (km)
                  'val': numpy.nan, 'idx': 4,
                  'short_name': 'barometric_altitude',
                  'units': 'km',
                  'long_name': 'Pressure altitude of GV (km)',
                  '_FillValue': "-99.999"},
                 'SRPALT':  {  # MTP Scan RMSE Pressure Alt (km)
                  'val': numpy.nan, 'idx': 5},
                 'SAAT':    {  # MTP Scan Avg Ambient Air Temp (deg_K)
                  'val': numpy.nan, 'idx': 6},
                 'SRAT':    {  # MTP Scan RMSE Ambient Air Temp (deg_K)
                  'val': numpy.nan, 'idx': 7},
                 'SALAT':   {  # MTP Scan Avg Latitude (degree_N)
                  'val': numpy.nan, 'idx': 8,
                  'short_name': 'latitude',
                  'units': 'degree_north',
                  'long_name': 'Latitude (deg)',
                  '_FillValue': "-999.99"},
                 'SRLAT':   {  # MTP Scan RMSE Latitude (degree_N)
                  'val': numpy.nan, 'idx': 9},
                 'SALON':   {  # MTP Scan Avg Longitude (degree_E)
                  'val': numpy.nan, 'idx': 10,
                  'short_name': 'longitude',
                  'units': 'degree_east',
                  'long_name': 'Longitude (deg)',
                  '_FillValue': "-9999.99"},
                 'SRLON':   {  # MTP Scan RMSE Longitude (degree_E)
                  'val': numpy.nan, 'idx': 11},
                 'SMCMD':   {  # MTP Scan Motor Commanded Position
                  'val': numpy.nan, 'idx': 12},
                 'SMENC':   {  # MTP Scan Motor Encoded Position
                  'val': numpy.nan, 'idx': 13},
                 },
    },
    'IWG1line': {
        # From project "ascii_parms" file
        're': "^IWG1,(........T......),(.*)",
        'found': False,
        'date': "",
        'asciiPacket': "",
        'data': [],
        'values': {'DATE': {  # IWG1 packet Date (YYYYMMDD)
                  'val': numpy.nan},
                 'TIME': {  # IWG1 packet Time (HHMMSS) converted to secs
                  'val': numpy.nan},
                 # Additional variables are created dynamically from list of
                 # vars read from ascii_parms
                 },
    },

    'Bline': {
        're': "(^B) (.*)",
        'found': False,
        'data': [],
        'values': {'SCNT': {  # MTP Scan Counts[Angle,Channel]
                  'val': [numpy.nan]*30,
                  'tb': [numpy.nan]*30}  # Calculated Brightness Temperatures
                 },
    },
    'M01line': {  # MTP Engineering Multiplxr
        're': "(^M01): (.*)",
        'found': False,
        'data': [],
        'values': {'VM08CNTE': {  # Voltage minus 8 Counts
                        'val': numpy.nan, 'idx': 0,
                        'fact': -2.73,  # factors from decodeM01 in VB6
                                        # used to convert counts to volts
                        'volts': numpy.nan,  # Calculated voltage
                        'name': "-8V  PS"},
                   'VVIDCNTE': {  # Vvid Counts
                        'val': numpy.nan, 'idx': 1,
                        'fact': 1,
                        'volts': numpy.nan,  # Calculated voltage
                        'name': "Video V."},
                   'VP08CNTE': {  # Voltage plus 8 Counts
                        'val': numpy.nan, 'idx': 2,
                        'fact': 2.78,
                        'volts': numpy.nan,  # Calculated voltage
                        'name': "+8V  PS"},
                   'VMTRCNTE': {  # Vmtr Count
                        'val': numpy.nan, 'idx': 3,
                        'fact': 7.79,
                        'volts': numpy.nan,  # Calculated voltage
                        'name': "+24V Step"},
                   'VSYNCNTE': {  # Vsyn Counts
                        'val': numpy.nan, 'idx': 4,
                        'fact': 7.79,
                        'volts': numpy.nan,  # Calculated voltage
                        'name': "+15V Syn"},
                   'VP15CNTE': {  # Volage plus 15 Counts
                        'val': numpy.nan, 'idx': 5,
                        'fact': 5.1,
                        'volts': numpy.nan,  # Calculated voltage
                        'name': "+15V PS"},
                   'VP05CNTE': {  # Voltage plus 5 Counts
                        'val': numpy.nan, 'idx': 6,
                        'fact': 2,
                        'volts': numpy.nan,  # Calculated voltage
                        'name': "VCC  PS"},
                   'VM15CNTE': {  # Voltage minus 15 Counts
                        'val': numpy.nan, 'idx': 7,
                        'fact': -5.1,
                        'volts': numpy.nan,  # Calculated voltage
                        'name': "-15V PS"},
                   },
    },

    'M02line': {  # MTP Engineering Multiplxr
        're': "(^M02): (.*)",
        'found': False,
        'data': [],
        'values': {'ACCPCNTE': {  # Acceler Counts
                        'val': numpy.nan, 'idx': 0,
                        'temperature': numpy.nan,  # Calculated acceleration
                        'name': "Acceler"},
                   'TDATCNTE': {  # T Data Counts
                        'val': numpy.nan, 'idx': 1,
                        'temperature': numpy.nan,  # Calculated temperature
                        'name': "T Data"},
                   'TMTRCNTE': {  # T Motor Counts
                        'val': numpy.nan, 'idx': 2,
                        'temperature': numpy.nan,  # Calculated temperature
                        'name': "T Motor"},
                   'TAIRCNTE': {  # T Pod Air Counts
                        'val': numpy.nan, 'idx': 3,
                        'temperature': numpy.nan,  # Calculated temperature
                        'name': "T Pod Air"},
                   'TSMPCNTE': {  # T Scan Counts
                        'val': numpy.nan, 'idx': 4,
                        'temperature': numpy.nan,  # Calculated temperature
                        'name': "T Scan"},
                   'TPSPCNTE': {  # T Power Supply Counts
                        'val': numpy.nan, 'idx': 5,
                        'temperature': numpy.nan,  # Calculated temperature
                        'name': "T Pwr Sup"},
                   'TNCCNTE': {  # T N/C Counts
                        'val': numpy.nan, 'idx': 6,
                        'temperature': numpy.nan,  # Calculated temperature
                        'name': "T N/C"},
                   'TSYNCNTE': {  # T Synth Counts
                        'val': numpy.nan, 'idx': 7,
                        'temperature': numpy.nan,  # Calculated temperature
                        'name': "T Synth"},
                   },
    },
    'Ptline': {  # MTP Platinum Multiplxr
        're': "(^Pt): (.*)",
        'found': False,
        'data': [],
        'values': {'TR350CNTP': {  # R350 Counts
                         'val': numpy.nan, 'idx': 0,
                         'resistance': numpy.nan,   # Calculated resistance
                         'temperature': numpy.nan,  # Calculated temperature
                         'name': "Rref 350"},       # Name displayed in GUI
                   'TTCNTRCNTP': {  # Target Center Temperature Counts
                         'val': numpy.nan, 'idx': 1,
                         'resistance': numpy.nan,   # Calculated resistance
                         'temperature': numpy.nan,  # Calculated temperature
                         'name': "Target 1"},       # Name displayed in GUI
                   'TTEDGCNTP': {  # Target Edge Temperature Counts
                         'val': numpy.nan, 'idx': 2,
                         'resistance': numpy.nan,   # Calculated resistance
                         'temperature': numpy.nan,  # Calculated temperature
                         'name': "Target 2"},       # Name displayed in GUI
                   'TWINCNTP': {  # Polyethelene Window Temperature Counts
                         'val': numpy.nan, 'idx': 3,
                         'resistance': numpy.nan,   # Calculated resistance
                         'temperature': numpy.nan,  # Calculated temperature
                         'name': "Window"},         # Name displayed in GUI
                   'TMIXCNTP': {  # Mixer Temperature Counts
                         'val': numpy.nan, 'idx': 4,
                         'resistance': numpy.nan,   # Calculated resistance
                         'temperature': numpy.nan,  # Calculated temperature
                         'name': "Mixer"},          # Name displayed in GUI
                   'TAMPCNTP': {  # Amplifier Temperature Counts
                         'val': numpy.nan, 'idx': 5,
                         'resistance': numpy.nan,   # Calculated resistance
                         'temperature': numpy.nan,  # Calculated temperature
                         'name': "Dblr Amp"},       # Name displayed in GUI
                   'TNDCNTP': {  # Noise Diode Temperature Counts
                         'val': numpy.nan, 'idx': 6,
                         'resistance': numpy.nan,   # Calculated resistance
                         'temperature': numpy.nan,  # Calculated temperature
                         'name': "Noise D."},       # Name displayed in GUI
                   'TR600CNTP': {  # R600 Counts
                         'val': numpy.nan, 'idx': 7,
                         'resistance': numpy.nan,   # Calculated resistance
                         'temperature': numpy.nan,  # Calculated temperature
                         'name': "Rref 600"},       # Name displayed in GUI
                   },
    },
    'Eline': {
        're': "(^E) (.*)",
        'found': False,
        'data': [],
        # The vector of counts produced by the MTP target. This vector is of
        # length 6 - three per channel for the noise diode turned on (stored as
        # the first three values) and three for it turned off (the next 3).
        'values': {'TCNT': {  # MTP Target Counts[Target,Channel]
                    'val': [numpy.nan]*6}
                   },
    },
    'tbi': "",  # Will hold an inverted brightness temperature array
    'BestWtdRCSet': "",  # Will hold an RC_Set_4Retrieval dictionary
    'ATP': "",  # Will hold an AtmosphericTemperatureProfile dictionary
}
